### PR TITLE
#1

### PR DESCRIPTION
**Pull request by robvenn**
Replaces readFileSync with async readFile

### CODE REVIEW

**Comment by phated** [comment]
I'm having a hard time reviewing this with the github diff tool.  I might need to pull it down locally and read just the end result.

**Comment by robvenn** [comment]
If it's too complex like this it could be refactored using promises or even async/await, as third-party modules for backwards compatibility.

**Comment by phated** [comment]
I don't think it's too complex, I just haven't had a chance to do a thorough review due to the github diff being bad.

Reminder: we want to keep ES5-compat so async/await would be out of the question.

**Comment by robvenn** [comment]
I see, didn't realise it was the diff, thought the code was not very clear. As far as compatibility I know this should be ES5 compatible that's why I was thinking about using [Bluebird](https://github.com/petkaantonov/bluebird) or [async-await](https://github.com/yortus/asyncawait) but I avoided adding additional dependencies for now. I do think it gets harder to reason about like this though.

**Comment by phated** [comment]
Assigned this to @contra incase he has time to review/make suggestions/help this week

**Comment by yocontra** [comment]
@robinvenneman Have you thought about using the `async` module (or smaller similar modules) to orchestrate the async stuff vs. doing it all by hand? I think it would clean this up a lot.

sidenote: index.js needs to be broken apart big time, not in this PR though. if you're interested, we'd love your help.

**Comment by robvenn** [comment]
@contra yes I did consider `async`, even mentioned it on #1 and then tried it without first. There's just so many libraries out there as well. If needed I can rework it in a few days.

And I agree about the index file, it should definitely be refactored. I'd be glad to help.

**Comment by robvenn** [comment]
Refactored the main methods with `async.waterfall`. It's a little cleaner, @contra what do you think?

File: index.js [hunk 1]
```diff
@@ -220,28 +145,110 @@ module.exports.add = function add (file, options, cb) {
...
+					if (ast.position) {
+						generator.addMapping({
+							original: ast.position.start,
+							generated: ast.position.start,
+							source: source,
+						});
+					}
+					for (var key in ast) {
+						if (key !== 'position') {
+							if (Object.prototype.toString.call(ast[key]) === '[object Object]') {
```
Comment (yocontra) [inline index.js]: Should use a cleaner way of doing this

File: index.js [hunk 2]
```diff
@@ -220,28 +145,110 @@ module.exports.add = function add (file, options, cb) {
...
+								registerTokens(ast[key]);
+							} else if (ast[key].constructor === Array) {
```
Comment (yocontra) [inline index.js]: Use Array.isArray

File: index.js [hunk 3]
```diff
@@ -220,28 +145,110 @@ module.exports.add = function add (file, options, cb) {
...
+								for (var i = 0; i < ast[key].length; i++) {
```
Comment (yocontra) [inline index.js]: Can you use functional iterators instead of forEach loops? Maps cleaner/less code.

This specific case would be one line: `ast[key].forEach(registerTokens)`

File: index.js [hunk 4]
```diff
@@ -220,28 +145,110 @@ module.exports.add = function add (file, options, cb) {
...
+					if (token.type.label === 'name') {
+						mapping.name = token.value;
+					}
+					generator.addMapping(mapping);
+				}
+				generator.setSourceContent(source, fileContent);
+				sourceMap = generator.toJSON();
+			} else if (fileType === '.css') {
+				var ast = css.parse(fileContent, { silent: true });
+				var registerTokens = function (ast) {
```
Comment (yocontra) [inline index.js]: Defining functions inside of if blocks is kind of sketch, should break these out to the highest scope wherever possible

**Review by yocontra — CHANGES_REQUESTED** [review]
Code needs a little cleanup, thanks for putting in the time!

File: index.js (hunk 3 above)
Comment (robvenn) [reply index.js]: Ok, I see what you mean, it's still not very functional with the side effects (generator, source) but used an IIFE to call the registerTokens first... Might need some more work to make it more pure.

File: index.js [hunk 5]
```diff
@@ -18,6 +18,29 @@ function unixStylePath (filePath) {
...
+		generator.addMapping({
+			original: ast.position.start,
+			generated: ast.position.start,
+			source: source
+		});
+	}
+	for (var key in ast) {
+		if (key === 'position' || !ast[key]) {
+			break;
+		} else {
```
Comment (yocontra) [inline index.js]: else not needed since it would `break` if the first condition is true

File: index.js [hunk 6]
```diff
@@ -18,6 +18,29 @@ function unixStylePath (filePath) {
...
+			if (ast[key] instanceof Array) {
+				for (var i = 0; i < ast[key].length; i++) {
```
Comment (yocontra) [inline index.js]: `ast[key].forEach()` instead of for loops

**Review by yocontra — DISMISSED** [review]

File: index.js (hunk 6 above)
Comment (robvenn) [reply index.js]: yeah this was fixed in the next commit, well spotted the unneeded `else` though.

**Comment by phated** [comment]
The level of nesting and amount of mutation in this code is really hard to follow.  I'm trying to split things out into top-level named functions to see how they are working and what they are mutating.   I've found a few things that should be easy to clean up but I'm going to keep digging.

**Comment by phated** [comment]
@robinvenneman it looks like tons of refactoring has been done in gulp-sourcemaps.  I'm going to experiment with moving that logic to this project.

**Comment by robvenn** [comment]
@phated ok, I recently noticed that too but didn't really follow the project much. If you can use any help just let me know.

File: lib/helpers.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  if (!source.map) {
+    return callback();
+  }
+
+  source.map.sourcesContent = source.map.sourcesContent || [];
+
+  // remove source map comment from source
+  file.contents = new Buffer(source.content, 'utf8');
+
+  async.forEachOf(source.map.sources, function(sourcePath, i, cb) {
```
Comment (phated) [inline lib/helpers.js]: This anonymous function can be split into a named function

**Comment by phated** [comment]
@robinvenneman I'm decided to be informed by it but change do the refactor myself. There are different ways that I'd like to tackle some stuff.

File: lib/helpers.js [hunk 8]
```diff
@@ -71,48 +71,61 @@ function fixImportedSourceMap(file, source, options, callback) {
...
+    if (source.map.sourcesContent[idx]) {
+      return cb();
     }
+
+    if (sourceRoot && isRemoteSource(sourceRoot)) {
+      assignSourcesContent(sourceContent, idx);
+      return cb();
+    }
+
+    var basePath = path.resolve(file.base, sourceRoot);
```
Comment (phated) [inline lib/helpers.js:102]: @robinvenneman I cleaned up this code (removed some branching logic) and got these tests passing.  If you understand this, please review and make sure I didn't inadvertently break something (I have no idea how thorough the tests are).

**Comment by phated** [comment]
@robinvenneman getting a lot closer.  Feel free to take a look when you get a chance.  I will be doing some more stuff tonight or tomorrow

**Comment by phated** [comment]
There's still a ton to do on this library, but most of the changes revolve around breaking old behavior/options/etc.  @robinvenneman @contra can you do a review on the point it's currently reached

I'd like to get this massive changeset merged and then start working on small, incremental changes.

File: lib/helpers.js (hunk 8 above)
Comment (robvenn) [reply lib/helpers.js:102]: @phated Afraid I can't really help much with that, also just refactored existing code and fixed the tests, they did look pretty thorough and added a few myself so hopefully it should be ok.

File: index.js [hunk 9]
```diff
@@ -1,184 +1,92 @@
...
-					sourceMap.sourcesContent[i] = sourceMap.sourcesContent[i] || null;
-					return;
-				}
-				var absPath = path.resolve(sourcePath, source);
-				sourceMap.sources[i] = unixStylePath(path.relative(file.base, absPath));
+/**
+ * Add a sourcemap to a vinyl file (async, with callback function)
+ * @param file
+ * @param options
+ * @param cb
```
Comment (robvenn) [inline index.js:13]: small mistake in the comment here, different name from the actual parameter

**Comment by phated** [comment]
I talked to @contra in IRC and he said this looks good.  Going to wrap up @robinvenneman's comment and ship this as a 0.1.0 "don't use this version".  Haha

**Comment by phated** [comment]
Oh, and I'm going to merge this without rebasing or squashing and do that later after we gut the stuff we don't want.
